File: src/test/js/formulas/math/UMINUSTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7Exception } from "../../../../main/js/errors/F7Exception";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { UMINUS } from "../../../../main/js/formulas/math/UMINUS";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("UMINUS", function () {
  it("should do normal operations", function () {
    assert.deepEqual(UMINUS.SELF.run(null, 10), -10);
    assert.deepEqual(UMINUS.SELF.run(null, -10), 10);
    assert.deepEqual(UMINUS.SELF.run(null, 0), 0);
    assert.deepEqual(UMINUS.SELF.run(null, 8278.28687), -8278.28687);
  });

  it("should do string conversion", function () {
    assert.deepEqual(UMINUS.SELF.run(null, "10"), -10);
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(
      (UMINUS.SELF.run(null, new ValueException()) as F7Exception).name,
      F7ExceptionName.VALUE
    );
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new UMINUS(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(2);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), -2);
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.equal(collateralLookup.callCount, 1);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, 2).add(0, 1, "A").build();
    assert.deepEqual(UMINUS.SELF.run(null, one), -2);
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual(
      (UMINUS.SELF.run(null, "A", "Too many") as NAException).name,
      F7ExceptionName.NA
    );
  });
});

File: src/test/java/io/protobase/f7/formulas/math/UMINUSTest.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class UMINUSTest extends TestFormula {
  @Test
  public void testApply() {
    assertThat(UMINUS.SELF.apply(null, 10.0)).isEqualTo(-10.0);
    assertThat(UMINUS.SELF.apply(null, -10.0)).isEqualTo(10.0);
    assertThat(UMINUS.SELF.apply(null, 0.0)).isEqualTo(0.0);
    assertThat(UMINUS.SELF.apply(null, 8278.28687)).isEqualTo(-8278.28687);
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(UMINUS.SELF.apply(null, "10.0")).isEqualTo(-10.0);
  }

  @Test
  public void testApply_boolean() {
    assertThat(UMINUS.SELF.apply(null, true)).isEqualTo(-1.0);
    assertThat(UMINUS.SELF.apply(null, false)).isEqualTo(0.0);
  }

  @Test
  public void testApply_grid() {
    assertThat(UMINUS.SELF.apply(null,
        Grid.builder().add(0, 0, -4.0)
            .add(0, 1, "Don't mind me.")
            .build()
    )).isEqualTo(4.0);
  }

  @Test
  public void test_lookup() {
    UMINUS F = new UMINUS(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(10.0);
    assertThat(F.apply(A1, M22_RANGE)).isEqualTo(-10.0);
    verify(collateralLookup).apply(A1, M22_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(UMINUS.SELF.apply(null, new ValueException())).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(UMINUS.SELF.apply(null, "A", "Too many")).isEqualTo(new NAException());
  }
}
